### ISSUE TITLE
bug - compile error - MICROPY_ENABLE_GC==0 && MICROPY_ENABLE_FINALISER==1

### ISSUE THREAD

**Issue by adritium** [micropython/micropython#3844]
https://github.com/micropython/micropython/blob/aace60a75e5afedadfc30d33e91ef7c6fa312405/py/malloc.c#L102-L107

If `MICROPY_ENABLE_GC==0`, should this code be allowed to be compiled even if `MICROPY_ENABLE_FINALISER==1`?

It causes a compile error.

**Comment by stinos** [comment]
What error exactly? The int to void* conversion?

**Comment by adritium** [comment]
`malloc.c(101,5): error : implicit declaration of function 'malloc_with_finaliser' [-Werror=implicit-function-declaration]`

**Comment by adritium** [comment]
I was implying - not strongly enough - that usage of finaliser macro should be tied to gc macro.

**Comment by dpgeorge** [comment]
@adritium do you have a suggestion what to do here, how to resolve the issue?

**Comment by adritium** [comment]
@dpgeorge 

> It's not really supported to have finilisation enabled with the GC disabled

Let's add `MICROPY_ENABLE_GC == 1` condition such that

```C
#if MICROPY_ENABLE_FINALISER &&  MICROPY_ENABLE_GC
 void *m_malloc_with_finaliser(size_t num_bytes) { 
     void *ptr = malloc_with_finaliser(num_bytes); 
     if (ptr == NULL && num_bytes != 0) { 
         m_malloc_fail(num_bytes); 
     }
```

**Comment by dpgeorge** [comment]
> Let's add MICROPY_ENABLE_GC == 1 condition such that

But that would just push the error message to a link error, like: `vfs_fat_file.c:(.text+0x6ba): undefined reference to m_malloc_with_finaliser`

Since it's not supported to do this I think it'd be better to emit macro error (using `#error`) if the GC is disabled with finalisers enabled.

**Comment by adritium** [comment]
What’s different between this issue and #3914 ?
In both there is a compile-time issue when one mpconfig.h macro is enabled and another one isn’t. 

Unless this one is different, I thought it’d make sense to have a consistent scheme: if a piece of code requires one or more mpconfig.h macros to properly compile/link then surround said code with ALL of them. 

But you know the code better; if this issue is an exception, please say why.

**Comment by dpgeorge** [comment]
Fixed by fcf621b066142cc221e5357f54e8156b1cb8c7fd
